ext/native/thin3d/VulkanRenderManager.cpp

```diff
@@ -0,0 +1,1328 @@
+				VkClearAttachment &attachment = attachments[numAttachments++];
+				attachment.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
+				attachment.colorAttachment = 0;
+				Uint8x4ToFloat4(attachment.clearValue.color.float32, c.clear.clearColor);
+			}
+			if (c.clear.clearMask & (VK_IMAGE_ASPECT_DEPTH_BIT | VK_IMAGE_ASPECT_STENCIL_BIT)) {
+				VkClearAttachment &attachment = attachments[numAttachments++];
+				attachment.aspectMask = 0;
+				if (c.clear.clearMask & VK_IMAGE_ASPECT_DEPTH_BIT) {
+					attachment.clearValue.depthStencil.depth = c.clear.clearZ;
+					attachment.aspectMask |= VK_IMAGE_ASPECT_DEPTH_BIT;
+				}
+				if (c.clear.clearMask & VK_IMAGE_ASPECT_STENCIL_BIT) {
+					attachment.clearValue.depthStencil.stencil = c.clear.clearStencil;
+					attachment.aspectMask |= VK_IMAGE_ASPECT_STENCIL_BIT;
+				}
+			}
+			if (numAttachments) {
+				vkCmdClearAttachments(cmd, numAttachments, attachments, 1, &rc);
+			}
+			break;
+		}
+		default:
+			ELOG("Unimpl queue command");
+			;
+		}
+	}
+	vkCmdEndRenderPass(cmd);
+
+	// Transition the framebuffer if requested.
+	if (fb && step.render.finalColorLayout != VK_IMAGE_LAYOUT_UNDEFINED) {
+		VkImageMemoryBarrier barrier{};
+		barrier.sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER;
+		barrier.oldLayout = fb->color.layout;
+		barrier.subresourceRange.layerCount = 1;
+		barrier.subresourceRange.levelCount = 1;
+		barrier.image = fb->color.image;
+		barrier.srcAccessMask = 0;
+		switch (barrier.oldLayout) {
+		case VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL:
+			barrier.srcAccessMask = VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT | VK_ACCESS_COLOR_ATTACHMENT_READ_BIT;
+			break;
+		case VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL:
+			barrier.srcAccessMask = VK_ACCESS_TRANSFER_WRITE_BIT;
+			break;
+		default:
+			Crash();
+		}
+		barrier.newLayout = step.render.finalColorLayout;
+		switch (barrier.newLayout) {
+		case VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL:
+			barrier.dstAccessMask = VK_ACCESS_SHADER_READ_BIT;
+			break;
+		default:
+			Crash();
+		}
+		barrier.subresourceRange.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
+
+		// we're between passes so it's OK.
+		// ARM Best Practices guide recommends these stage bits.
+		vkCmdPipelineBarrier(cmd, VK_PIPELINE_STAGE_ALL_GRAPHICS_BIT, VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT, 0, 0, nullptr, 0, nullptr, 1, &barrier);
+		fb->color.layout = barrier.newLayout;
+	}
+}
+
+void VulkanRenderManager::PerformBindFramebufferAsRenderTarget(const VKRStep &step, VkCommandBuffer cmd, int swapChainImage) {
+	VkFramebuffer framebuf;
+	int w;
+	int h;
+	VkImageLayout prevLayout;
+	if (step.render.framebuffer) {
+		VKRFramebuffer *fb = step.render.framebuffer;
+		framebuf = fb->framebuf;
+		w = fb->width;
+		h = fb->height;
+		prevLayout = fb->color.layout;
+	} else {
+		framebuf = framebuffers_[swapChainImage];
+		w = vulkan_->GetBackbufferWidth();
+		h = vulkan_->GetBackbufferHeight();
+	}
+
+#if 0
+	// This part is based on faulty old thinking.
```

Comment:
Should we just delete?

I'm missing what's faulty - although this is definitely suboptimal, and if we're (hopefully) optimizing out clears, it shouldn't be needed as often?

-[Unknown]